### DISCUSSION TITLE
How to use async database connections and share it in request.app.state

### DISCUSSION THREAD

**Discussion by harshguptaserver** [fastapi/fastapi#11534] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from contextlib import asynccontextmanager

@asynccontextmanager
async def setup_connection_pools(app:FastAPI):
    print("isinstance(app.state, State)")
    print(isinstance(app.state, State))
    my_redis_conn = await initialize_cache()
    app.state.redis = my_redis_conn
    yield

app = FastAPI(lifespan=setup_connection_pools)

@app.post("/debug")
async def test_debug(request: Request):
  print(request.app.state.redis)
```

### Description

Requirement: I want to create database connection pools after deprecation on app.on_event("startup") and use request.app.state.connection_pool_object

I was using :
```
@app.on_event("startup")
async def create_redis():
    # :ref: https://gist.github.com/h1code2/f3bba962a694c126ae320a5d87cd5875
    app.state.redis = await initialize_cache()
    app.state.mongodb_client, app.state.mongodb_vault = await get_mongodb_connection()
```
Then in subrouter i used to use 
```
vault_api = APIRouter()

@vault_api.post("/action/create/{a_vault_name:str}/{object_type:int}")
async def add_vault_item(input_vault_data: CreateVaultItem, request: Request, a_vault_name: str,
                         object_type: int = 0) -> JSONResponse:
..
     res_vault, err_vault = await db_storage_get_a_dj_vault(
        cid=dj_id,
        vault_dj_name=a_vault_name,
        vault_collection=request.app.state.mongodb_vault
    )
```

Now I upgraded to latest fastapi version and I am unable to use on event startup, I tried referring to lifespan page but it does not help in subrouter.

### Operating System

Linux

### Operating System Details

cat /etc/os-release
NAME="Pop!_OS"
VERSION="22.04 LTS"
ID=pop
ID_LIKE="ubuntu debian"
PRETTY_NAME="Pop!_OS 22.04 LTS"
VERSION_ID="22.04"

### FastAPI Version

0.111.0

### Pydantic Version

2.7.1

### Python Version

Python 3.12.2

### Additional Context

I want to be able to use request.app.state.x again using lifespan or any other way since on_startup is deprecated.

**Comment by harshguptaserver** [comment]
The only way I found is :ref: 

https://stackoverflow.com/questions/76322463/how-to-initialise-a-global-object-or-variable-and-reuse-it-in-every-fastapi-endp

**Comment by hasansezertasan** [comment]
You can also take a look at middlewares.

Starlette Admin has a middleware for SQLAlchemy Contrib, check that out.

**Comment by YuriiMotov** [comment, marked as answer]
You can use the following approach:

```python
from contextlib import asynccontextmanager
from typing import Annotated

from fastapi import Depends, FastAPI, Request


async def initialize_cache():
    return "fake conn"


@asynccontextmanager
async def setup_connection_pools(app:FastAPI):
    my_redis_conn = await initialize_cache()
    yield {"redis_conn": my_redis_conn}

app = FastAPI(lifespan=setup_connection_pools)


async def get_redis_conn(request: Request):
    redis_conn = getattr(request.state, "redis_conn", None)
    if redis_conn is None:
        raise RuntimeError("Redis connection not initialized")
    return redis_conn


@app.post("/debug")
async def test_debug(redis_conn: Annotated[str, Depends(get_redis_conn)]):
    print(redis_conn)

```
